consensus/core/src/linearizer.rs

```diff
@@ -120,42 +149,82 @@ impl Linearizer {
-                            // We skip the block if we already committed it or we reached a
-                            // round that we already committed.
-                            // TODO: for Fast Path we need to ammend the recursion rule here and allow us to commit blocks all the way up to the `gc_round`.
-                            // Some additional work will be needed to make sure that we keep the uncommitted blocks up to the `gc_round` across commits.
-                            !committed.contains(ancestor)
-                                && last_committed_rounds[ancestor.author] < ancestor.round
-                        })
-                        .filter(|ancestor| {
-                            // Keep the block if GC is not enabled or it is enabled and the block is above the gc_round. We do this
-                            // to stop the recursion early and avoid going to deep when it's unnecessary.
-                            !gc_enabled || ancestor.round > gc_round
-                        })
-                        .collect::<Vec<_>>(),
-                )
-                .into_iter()
-                .map(|ancestor_opt| {
-                    ancestor_opt.expect("We should have all uncommitted blocks in dag state.")
-                })
-                .collect();
-
-            for ancestor in ancestors {
-                buffer.push(ancestor.clone());
-                assert!(committed.insert(ancestor.reference()));
+
+        // The new logic will perform the recursion without stopping at the highest round round that has been committed per authority. Instead it will
+        // allow to commit blocks that are lower than the highest committed round for an authority but higher than gc_round.
+        if context.protocol_config.consensus_linearize_subdag_v2() {
+            assert!(
+                dag_state.set_committed(&leader_block_ref),
+                "Leader block with reference {:?} attempted to be committed twice",
+                leader_block_ref
+            );
+
+            while let Some(x) = buffer.pop() {
+                to_commit.push(x.clone());
+
+                let ancestors: Vec<VerifiedBlock> = dag_state
+                    .get_blocks(
+                        &x.ancestors()
+                            .iter()
+                            .copied()
+                            .filter(|ancestor| {
+                                ancestor.round > gc_round && !dag_state.is_committed(ancestor)
+                            })
+                            .collect::<Vec<_>>(),
+                    )
+                    .into_iter()
+                    .map(|ancestor_opt| {
+                        ancestor_opt.expect("We should have all uncommitted blocks in dag state.")
+                    })
+                    .collect();
+
+                for ancestor in ancestors {
+                    buffer.push(ancestor.clone());
+                    assert!(
```

Comment:
Why not combine the `dag_state.set_committed(&ancestor.reference())` asserts and call it after `to_commit.push(x.clone());`?

Reply:
Could you please clarify a little bit more on this? Do you suggest moving the `dag_state.set_committed(&ancestor.reference())` after the `to_commit.push(x.clone());` ? That would not give desired results given that we do BFS here. I mean, if we don't mark the blocks as committed here then we'll revisit them and try to re-commit them. Unless I am totally misunderstanding what you mean.

Reply:
Sounds good.